File: Algorithm/Graph/DFS_BFS/Programmers/43162_네트워크_2.py
# https://programmers.co.kr/learn/courses/30/lessons/43162
# 프로그래머스 #43162
# 네트워크

#%%
n = 3
computers = [[1, 1, 0], [1, 1, 0], [0, 0, 1]]
# computers = [[1, 1, 0], [1, 1, 1], [0, 1, 1]]
# %%
def dfs(index, visited):
    print('_______________')
    print(index, visited)
    for i, connection in enumerate(computers[index]):
        print(f'check connection : {i}, {connection}')
        if i not in visited and connection == 1:
            visited.append(i)
            print(f'add node {i}, visited : {visited}')
            dfs(i, visited)
    
#%%
visited = []
cnt = 0
# %%
# i : 0, 1, 2
for index in range(len(computers)):
    if index not in visited:
        visited.append(index)
        dfs(index, visited)
        cnt += 1
# %%
cnt
# %%
def solution(n, computers):
    def dfs(index, visited):
        for i, connection in enumerate(computers[index]):
            if i not in visited and connection == 1:
                visited.append(i)
                dfs(i, visited)
                
    visited = []
    cnt = 0 

    for index in range(len(computers)):
        if index not in visited:
            visited.append(index)
            dfs(index, visited)
            cnt += 1
    return cnt

File: Algorithm/Graph/DFS_BFS/Programmers/43162_네트워크_2.go
// https://programmers.co.kr/learn/courses/30/lessons/43162
// 프로그래머스 #43162
// 네트워크

package main

import "fmt"

var n int
var computers [][]int

func contains(n int, slice []int) bool {
	for _, v := range slice {
		if v == n {
			return true
		}
	}
	return false
}

func dfs(index int, visited *[]int) {
	for i, connection := range computers[index] {
		if !contains(i, *visited) && connection == 1 {
			*visited = append(*visited, i)
			dfs(i, visited)
		}
	}
}

func solution(n int, computers [][]int) int {
	count := 0
	var visited []int

	for i := 0; i < len(computers); i++ {
		if visited == nil || !contains(i, visited) {
			visited = append(visited, i)
			dfs(i, &visited)
			count += 1
		}
	}
	fmt.Printf("count : %d", count)
	return count
}

func main() {
	n = 3
	computers = [][]int{
		{1, 1, 0},
		{1, 1, 0},
		{0, 0, 1},
	}
	// computers = [][]int{
	// {1, 1, 0},
	// {1, 1, 1},
	// {0, 1, 1},
	// }
	solution(n, computers)
}

func contains(n int, slice []int) bool {
	if slice == nil {
		return false
	}
	for _, v := range slice {
		if v == n {
			return true
		}
	}
	return false
}

// solution
// func dfs(index int, visited *[]int, computers *[][]int) {
//     for i, connection := range (*computers)[index] {
// 		if !contains(i, *visited) && connection == 1 {
// 			*visited = append(*visited, i)
// 			dfs(i, visited, computers)
// 		}
// 	}
// }

// func solution(n int, computers [][]int) int {

// 	count := 0
// 	var visited []int

// 	for i := 0; i < len(computers); i++ {
// 		if visited == nil || !contains(i, visited) {
// 			visited = append(visited, i)
// 			dfs(i, &visited, &computers)
// 			count += 1
// 		}
// 	}
// 	return count
// }
